napari/utils/tree/_tests/test_tree_model.py

```diff
@@ -0,0 +1,182 @@
+    with pytest.raises(IndexError) as e:
+        g1_1_0.unparent()
+    assert "Cannot unparent orphaned Node" in str(e)
+
+
+def test_traverse(tree):
+    """Test depth first traversal."""
+    # iterating a group just returns its children
+    assert [x.name for x in tree] == ['1', 'g1', '8', '9']
+    # traversing a group does a depth first traversal, including both groups
+    # and nodes
+    names = [x.name for x in tree.traverse()]
+    e = ['root', '1', 'g1', '2', 'g2', '3', '4', '5', '6', '7', '8', '9']
+    assert names == e
+
+    # traversing leaves_only=True returns only the Nodes, not the Groups
+    names = [x.name for x in tree.traverse(leaves_only=True)]
+    e = ['1', '2', '3', '4', '5', '6', '7', '8', '9']
+    assert names == e
+
+    assert tree.is_group()
+    g1 = tree[1]
+    assert g1.parent is tree
+    assert g1.name == 'g1' and g1.is_group()
+    g2 = g1[1]
+    assert g2.parent is g1
+    assert g2.name == 'g2' and g2.is_group()
+
+
+def test_slicing(tree):
+    """Indexing into a group returns a group instance."""
+    assert tree.is_group()
+    slc = tree[::-2]
+    assert slc.is_group()
+    expected = ['Group', '9', 'g1', '2', 'g2', '3', '4', '5', '6', '7']
+    assert [x.name for x in slc.traverse()] == expected
+
+
+def test_contains(tree):
+    """Test that the ``in`` operator works for nested nodes."""
+    g1 = tree[1]
+    assert g1.name == 'g1'
+    assert g1 in tree
+
+    g1_0 = g1[0]
+    assert g1_0.name == '2'
+    assert g1_0 in g1
+    assert g1_0 in tree
+
+    # If you need to know if an item is an immediate child, you can use index
+    assert tree.index(g1) == 1
+    with pytest.raises(ValueError) as e:
+        tree.index(g1_0)
+    assert "is not in list" in str(e)
+
+    g2 = g1[1]
+    assert g2.name == 'g2'
+    assert g2.is_group()
+    assert g2 in tree
+
+    g2_0 = g2[0]
+    assert g2_0.name == '3'
+
+
+def test_deletion(tree):
+    """Test that deletion removes parent"""
+    g1 = tree[1]  # first group in tree
+    assert g1.parent is tree
+    assert g1 in tree
+    n1 = g1[0]  # first item in group1
+
+    del tree[1]  # delete g1 from the tree
+    assert g1.parent is not tree
+    assert g1 not in tree
+    # the tree no longer has g1 or any of its children
+    assert [x.name for x in tree.traverse()] == ['root', '1', '8', '9']
+
+    # g1 remains intact
+    expected = ['g1', '2', 'g2', '3', '4', '5', '6', '7']
+    assert [x.name for x in g1.traverse()] == expected
+    expected = ['2', '3', '4', '5', '6', '7']
+    assert [x.name for x in g1.traverse(leaves_only=True)] == expected
+
+    # we can also delete slices, including extended slices
+    del g1[1::2]
+    assert n1.parent is g1  # the g1 tree is still intact
+    assert [x.name for x in g1.traverse()] == ['g1', '2', '5', '7']
+
+
+def test_nested_deletion(tree):
+    """Test that we can delete nested indices from the root."""
+    # a tree is a NestedEventedList, so we can use nested_indices
+    node5 = tree[1, 2]
```

Comment:
> but a dedicated test which was something like `test_nested_indexing`

how about that actual test :joy:  (it's in master already)

https://github.com/napari/napari/blob/a91ba740d767feab9f6f3135bcb34489901ebaa5/napari/utils/events/_tests/test_evented_list.py#L234-L235

> but I see the Tree as something we'll want to engage with a lot so it might be nice to have a test for that here too

hmmm, but all of the logic and methods being tested are over in `_nested_list.py`


> Also can the tree be indexed by name, so assert tree['g1', '5'].name == '5'

oh man :joy: ... Yes, you can search nested lists by name [see `test_nested_custom_lookup`](https://github.com/napari/napari/pull/2217/files#diff-25f46b91b3c6443994ba1261120fb13bb2fd9a682300bd96a64d2d20dc763ce5R138) ... but I don't think you can also do it with a tuple like that.  so, you'd have to do `tree['g1']['5']` ... that one would be a bit tough

Reply:
ok yeah, that's fair maybe not need to duplicate test if literally would be identical and no reason to suspect that any further additions to `Group` would break it.

>  so, you'd have to do tree['g1']['5'] ... that one would be a bit tough

ok, makes sense - maybe one day!